src/validation.cpp

```diff
@@ -1060,28 +1176,93 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
     // checks pass, to mitigate CPU exhaustion denial-of-service attacks.
     PrecomputedTransactionData txdata;
 
-    if (!PolicyScriptChecks(args, ws, txdata)) return MempoolAcceptResult(ws.m_state);
+    if (!PolicyScriptChecks(args, ws, txdata)) return MempoolAcceptResult(*ws.m_ptx, ws.m_state);
 
-    if (!ConsensusScriptChecks(args, ws, txdata)) return MempoolAcceptResult(ws.m_state);
+    if (!ConsensusScriptChecks(args, ws, txdata)) return MempoolAcceptResult(*ws.m_ptx, ws.m_state);
 
     // Tx was accepted, but not added
     if (args.m_test_accept) {
-        return MempoolAcceptResult(std::move(ws.m_replaced_transactions), ws.m_base_fees);
+        return MempoolAcceptResult(*ws.m_ptx, std::move(ws.m_replaced_transactions), ws.m_base_fees);
     }
 
-    if (!Finalize(args, ws)) return MempoolAcceptResult(ws.m_state);
+    if (!Finalize(args, ws)) return MempoolAcceptResult(*ws.m_ptx, ws.m_state);
 
     GetMainSignals().TransactionAddedToMempool(ptx, m_pool.GetAndIncrementSequence());
 
-    return MempoolAcceptResult(std::move(ws.m_replaced_transactions), ws.m_base_fees);
+    return MempoolAcceptResult(*ws.m_ptx, std::move(ws.m_replaced_transactions), ws.m_base_fees);
+}
+
+std::vector<MempoolAcceptResult> MemPoolAccept::AcceptMultipleTransactions(std::vector<CTransactionRef>& txns, ATMPArgs& args)
+{
+    AssertLockHeld(cs_main);
+    std::vector<Workspace> workspaces{};
+    const int package_size = txns.size();
+    workspaces.reserve(package_size);
+    std::transform(txns.begin(), txns.end(), std::back_inserter(workspaces), [](CTransactionRef& tx) {
+        return Workspace(tx);
+    });
+    std::vector<MempoolAcceptResult> results;
+    results.reserve(package_size);
+
+    LOCK(m_pool.cs);
+    // Do all PreChecks first and fail fast to avoid running expensive script checks when unnecessary.
+    for (Workspace& ws : workspaces) {
+        if (!PreChecks(args, ws)) {
+            // Exit early to avoid doing pointless work. Return results in the same order as input txns.
+            const auto failed_or_unfinished = [&, failed_ptx = ws.m_ptx](Workspace& ws) {
+                                              return ws.m_ptx == failed_ptx
+                                              ? MempoolAcceptResult(*ws.m_ptx, ws.m_state, /* finished */ true)
+                                              : MempoolAcceptResult(*ws.m_ptx, ws.m_state, /* finished */ false);
+            };
+            std::transform(workspaces.begin(), workspaces.end(), std::back_inserter(results), failed_or_unfinished);
+            return results;
+        }
+        m_view.PackageAddTransaction(ws.m_ptx);
+    }
+
+    // Now that we have verified all inputs are available and there are no conflicts in the package,
+    // clear the temporary coins (m_temp_added and m_temp_spent), otherwise script checks will error
+    // on coins that are spent within the package.
+    m_view.ClearTemporaryCoins();
+
+    for (Workspace& ws : workspaces) {
```

Comment:
I love `std::transform` more than most, but I think this would be simpler if you just pushed the results onto `results` as you went (and then filled in the remaining `results` as `UNFINISHED` if you fail on any tx).